src/client/texturesource.cpp

```diff
@@ -1848,58 +1856,115 @@ bool TextureSource::generateImagePart(std::string_view part_of_name,
+/// A helper function for blit_with_alpha to support different endianesses
+template<bool overlay, class Colour>
+void blit_with_alpha_any_endian(video::IImage *src, video::IImage *dst,
+	const v2s32 &dst_pos, const v2u32 &size)
 {
-	auto src_dim = src->getDimension();
-	auto dst_dim = dst->getDimension();
+	if (dst->getColorFormat() != video::ECF_A8R8G8B8)
+		throw BaseException("blit_with_alpha() supports only ECF_A8R8G8B8 "
+			"destination images.");
+
+	auto src_dim{src->getDimension()};
+	auto dst_dim{dst->getDimension()};
+	bool drop_src{false};
+	if (src->getColorFormat() != video::ECF_A8R8G8B8) {
+		video::IVideoDriver *driver{RenderingEngine::get_video_driver()};
+		video::IImage *src_converted{driver->createImage(video::ECF_A8R8G8B8,
+			src_dim)};
+		if (!src_converted)
+			throw BaseException("blit_with_alpha() failed to convert the "
+				"source image to ECF_A8R8G8B8.");
+		src->copyTo(src_converted);
+		src = src_converted;
+		drop_src = true;
+	}
+
+	Colour *pixels_src{reinterpret_cast<Colour *>(src->getData())};
+	Colour *pixels_dst{reinterpret_cast<Colour *>(dst->getData())};
 
 	// Limit y and x to the overlapping ranges
 	// s.t. the positions are all in bounds after offsetting.
-	for (u32 y0 = std::max(0, -dst_pos.Y);
-			y0 < std::min<s64>({size.Y, src_dim.Height, dst_dim.Height - (s64) dst_pos.Y});
-			++y0)
-	for (u32 x0 = std::max(0, -dst_pos.X);
-			x0 < std::min<s64>({size.X, src_dim.Width, dst_dim.Width - (s64) dst_pos.X});
-			++x0)
-	{
-		s32 src_x = src_pos.X + x0;
-		s32 src_y = src_pos.Y + y0;
-		s32 dst_x = dst_pos.X + x0;
-		s32 dst_y = dst_pos.Y + y0;
-		video::SColor src_c = src->getPixel(src_x, src_y);
-		video::SColor dst_c = dst->getPixel(dst_x, dst_y);
-		if (!overlay || (dst_c.getAlpha() == 255 && src_c.getAlpha() != 0)) {
-			dst_c = blitPixel(src_c, dst_c, src_c.getAlpha());
-			dst->setPixel(dst_x, dst_y, dst_c);
+	u32 x_start{static_cast<u32>(std::max(0, -dst_pos.X))};
+	u32 y_start{static_cast<u32>(std::max(0, -dst_pos.Y))};
+	u32 x_end{static_cast<u32>(std::min<s64>({size.X, src_dim.Width,
+		dst_dim.Width - (s64) dst_pos.X}))};
+	u32 y_end{static_cast<u32>(std::min<s64>({size.Y, src_dim.Height,
+		dst_dim.Height - (s64) dst_pos.Y}))};
+	for (u32 y0{y_start}; y0 < y_end; ++y0) {
+		size_t i_src{y0 * src_dim.Width + x_start};
+		size_t i_dst{(dst_pos.Y + y0) * dst_dim.Width
+			+ dst_pos.X + x_start};
+		for (u32 x0{x_start}; x0 < x_end; ++x0) {
+			blit_pixel<overlay, Colour>(pixels_src[i_src++],
+				pixels_dst[i_dst++]);
 		}
 	}
+	if (drop_src)
+		src->drop();
+}
+
+}  // namespace
+
+template<bool overlay>
+static void blit_with_alpha(video::IImage *src, video::IImage *dst,
+	const v2s32 &dst_pos, const v2u32 &size)
+{
+	u32 one{1};
+	bool is_little_endian{*reinterpret_cast<u8 *>(&one) == 1};
+	if (is_little_endian) {
+		struct Colour {
+			u8 b, g, r, a;
+		};
+		blit_with_alpha_any_endian<overlay, Colour>(src, dst, dst_pos, size);
+	} else {
+		struct Colour {
+			u8 a, r, g, b;
+		};
```

Comment:
do we need to add `__attribute__((packed))` here or does the C++ standard guarantee no padding?

maybe an approach like [here](https://github.com/minetest/minetest/blob/master/src/client/imagefilters.cpp#L74) would be better where data is read as u32 and then casted to `SColor`
(though it looks like we forgot to handle endian there)

Reply:
It does not guarantee packing (at least not in general) so packing would be needed. The latest code doesn't reinterpret_cast u32 to a color struct anymore.

> though it looks like we forgot to handle endian there

SColor just wraps u32 and the methods to get color values calculate the u8 values each time they are executed, so I think endianess does not need to be handled there.